### ISSUE TITLE
Not compatible defined variable in meta dependencies

### ISSUE THREAD

**Issue by iwai** [ansible/ansible#14840]
this is bug?? or change specification?
##### Issue Type:
- Bug Report

In meta

```
dependencies:
  - { role: "php-phalcon", pkg_name: "php-phalcon2" }
  - { role: "nginx" }
```

In defaults/main.yml for php-phalcon role 

```

---
  pkg_name:  php-phalcon
```

In defaults/main.yml for nginx role 

```

---
  pkg_name:  nginx
```
##### Ansible Version:
- ansible version 1.9.4
- ansible version 2.0.0

```
# cat /etc/system-release
CentOS release 6.7 (Final)
# python --version
Python 2.6.6
```
##### actual result ansible version 1.9.4

```
TASK: [nginx | debug var=pkg_name] ************************* 
ok: [127.0.0.1] => {
    "var": {
        "pkg_name": "nginx"
    }
}
```
##### actual result ansible version 2.0.0

```
TASK: [nginx | debug var=pkg_name] ************************* 
ok: [127.0.0.1] => {
    "var": {
        "pkg_name": "php-phalcon2"
    }
}
```

i wish expected same result ansible version 2.0.0 and 1.9.4

**Comment by jimi-c** [comment]
@iwai can you please include a full example to reproduce this?

**Comment by benjifisher** [comment]
This looks to me like the same problem as Issue #14411.  I ran into the same thing:  when the role `foo` is declared as a dependency of the role `bar`, then variables passed to `foo` in `bar/meta/main.yml` override those defined in `foo/defaults/main.yml` even if `foo` is included before `bar`.  Here is my example:

**playbook.yml**

``` yaml
- hosts: local

  roles:
    - { role: foo, foovar: "Kilroy was here!" }
    - foo
    - bar
```

**hosts**

```
[local]
localhost
```

**roles/foo/defaults/main.yml**

``` yaml
---
foovar: "unchanged"
```

**roles/foo/meta/main.yml**

``` yaml
---
allow_duplicates: yes
```

**roles/foo/tasks/main.yml**

``` yaml
---
- name: "Foo: what is going on?"
  debug: var=foovar
```

**roles/bar/meta/main.yml**

``` yaml
---
dependencies:
  - { role: foo, foovar: "Balrog was here!" }
```

**roles/bar/tasks/main.yml**

``` yaml
---
- name: "Bar--here I am"
  debug: msg="Bar--here I am"
```

Here is the result of running the playbook:

```
$ ansible-playbook --version
ansible-playbook 2.0.1.0
  config file = /Users/bfisher/.ansible.cfg
  configured module search path = Default w/o overrides
$ ansible-playbook playbook.yml -i hosts

PLAY ***************************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [foo : Foo: what is going on?] ********************************************
ok: [localhost] => {
    "foovar": "Kilroy was here!"
}

TASK [foo : Foo: what is going on?] ********************************************
ok: [localhost] => {
    "foovar": "Balrog was here!"
}

TASK [foo : Foo: what is going on?] ********************************************
ok: [localhost] => {
    "foovar": "Balrog was here!"
}

TASK [bar : Bar--here I am] ****************************************************
ok: [localhost] => {
    "msg": "Bar--here I am"
}

PLAY RECAP *********************************************************************
localhost                  : ok=5    changed=0    unreachable=0    failed=0   
```

If I remove `bar` from the playbook, then I get this, so variables set when `foo` is included directly by the playbook to **not** override the defaults:

```
$ ansible-playbook playbook.yml -i hosts

PLAY ***************************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [foo : Foo: what is going on?] ********************************************
ok: [localhost] => {
    "foovar": "Kilroy was here!"
}

TASK [foo : Foo: what is going on?] ********************************************
ok: [localhost] => {
    "foovar": "unchanged"
}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0   
```

**Comment by benjifisher** [comment]
@jimi-c :  Based on my earlier comment, can you remove the "needs_info" label?

**Comment by jimi-c** [comment]
Running on both devel and stable-2.1, it appears this issue has been fixed:

<pre>
# ansible-playbook -vv -i hosts test.yml 
Using /etc/ansible/ansible.cfg as config file
PLAYBOOK: test.yml *************************************************************
1 plays in test.yml
PLAY [local] *******************************************************************
TASK [setup] *******************************************************************
ok: [localhost]
TASK [foo : Foo: what is going on?] ********************************************
task path: /root/testing/14840/roles/foo/tasks/main.yml:3
ok: [localhost] => {
    "foovar": "Kilroy was here!"
}
TASK [foo : Foo: what is going on?] ********************************************
task path: /root/testing/14840/roles/foo/tasks/main.yml:3
ok: [localhost] => {
    "foovar": "unchanged"
}
TASK [foo : Foo: what is going on?] ********************************************
task path: /root/testing/14840/roles/foo/tasks/main.yml:3
ok: [localhost] => {
    "foovar": "Balrog was here!"
}
TASK [bar : Bar--here I am] ****************************************************
task path: /root/testing/14840/roles/bar/tasks/main.yml:3
ok: [localhost] => {
    "msg": "Bar--here I am"
}
PLAY RECAP *********************************************************************
localhost                  : ok=5    changed=0    unreachable=0    failed=0   
</pre>

As such, this fix will be included in the 2.1 release, which should be out soon. If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
